src/Mvc/Mvc.Core/src/ModelBinding/Metadata/JsonMetadataProvider.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Metadata;
+
+/// <summary>
+/// An implementation of <see cref="IDisplayMetadataProvider"/> and <see cref="IValidationMetadataProvider"/> for
+/// the System.Text.Json.Serialization attribute classes.
+/// </summary>
+internal class JsonMetadataProvider : IDisplayMetadataProvider, IValidationMetadataProvider
+{
+    private readonly JsonNamingPolicy? _jsonNamingPolicy;
+
+    /// <summary>
+    /// Creates a new <see cref="JsonMetadataProvider"/>.
+    /// </summary>
+    public JsonMetadataProvider()
+        : this(JsonNamingPolicy.CamelCase)
+    {
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="JsonMetadataProvider"/> with an optional <see cref="JsonNamingPolicy"/>
+    /// </summary>
+    /// <param name="jsonNamingPolicy">The <see cref="JsonNamingPolicy"/> to be used to convert the property name</param>
+    public JsonMetadataProvider(JsonNamingPolicy? jsonNamingPolicy)
+        => _jsonNamingPolicy = jsonNamingPolicy;
+
+    /// <inheritdoc />
+    public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
+    {
+        if (context == null)
+        {
+            throw new ArgumentNullException(nameof(context));
+        }
+
+        var propertyName = ReadPropertyNameFrom(context.Attributes);
+
+        if (!string.IsNullOrEmpty(propertyName))
+        {
+            context.DisplayMetadata.DisplayName = () => propertyName;
```

Comment:
Should this fall back to `context.Key.Name` like you do for `ValidationModelName`?

Reply:
Good catch! That is wrong for sure.

Reply:
Actually not! In this case if the property is not set, we will use the display name function already available. That is why the condition is `!string.IsNullOrEmpty(propertyName)`

Reply:
In the `ValidationModelName` I would look to make sure the casing is correct, so, I will use the `context.Key.Name` to convert to the right case when the property attribute is not there.